File: core/git_context.py
"""Git context reader with parallel subprocess execution and silent fallback."""
import subprocess
import re
from concurrent.futures import ThreadPoolExecutor, as_completed

SUBPROCESS_TIMEOUT = 5

def load_git_context(cwd):
    """Load git context from the given directory using 4 parallel subprocesses.

    Returns dict with commits_today, diff_lines, first_commit_time, repo_path.
    Any subprocess failure silently falls back to safe defaults.
    """
    commands = {
        'commits_today': ['git', 'log', '--oneline', '--since=midnight'],
        'diff_lines': ['git', 'diff', '--stat', 'HEAD'],
        'first_commit_time': ['git', 'log', '--format=%ai', '--since=midnight'],
        'repo_path': ['git', 'rev-parse', '--show-toplevel']
    }

    results = {
        'commits_today': 0,
        'diff_lines': 0,
        'first_commit_time': None,
        'repo_path': None
    }

    with ThreadPoolExecutor(max_workers=4) as executor:
        futures = {executor.submit(_run_git_cmd, cmd, cwd): key for key, cmd in commands.items()}

        for future in as_completed(futures):
            key = futures[future]
            try:
                output = future.result()
                if output is not None:
                    if key == 'commits_today':
                        results[key] = _parse_commits_today(output)
                    elif key == 'diff_lines':
                        results[key] = _parse_diff_lines(output)
                    elif key == 'first_commit_time':
                        results[key] = _parse_first_commit_time(output)
                    elif key == 'repo_path':
                        results[key] = _parse_repo_path(output)
            except Exception:
                pass

    return results

def _run_git_cmd(cmd, cwd):
    """Run a git command and return stdout, or None on any failure."""
    try:
        result = subprocess.run(
            cmd,
            cwd=cwd,
            capture_output=True,
            text=True,
            timeout=SUBPROCESS_TIMEOUT,
            check=False
        )
        if result.returncode == 0:
            return result.stdout
        return None
    except (subprocess.TimeoutExpired, FileNotFoundError, Exception):
        return None

def _parse_commits_today(output):
    """Count lines in git log output."""
    lines = [line for line in output.strip().split('\n') if line]
    return len(lines)

def _parse_diff_lines(output):
    """Extract insertions + deletions from git diff --stat output."""
    if not output.strip():
        return 0

    insertions = 0
    deletions = 0

    match = re.search(r'(\d+) insertion', output)
    if match:
        insertions = int(match.group(1))

    match = re.search(r'(\d+) deletion', output)
    if match:
        deletions = int(match.group(1))

    return insertions + deletions

def _parse_first_commit_time(output):
    """Extract earliest commit time (last line) from git log output."""
    lines = [line for line in output.strip().split('\n') if line]
    if not lines:
        return None
    return lines[-1]

def _parse_repo_path(output):
    """Extract repo path from git rev-parse output."""
    path = output.strip()
    return path if path else None

File: src/core/gitContext.ts
import { execFile } from 'child_process'
import { promisify } from 'util'

const execFileAsync = promisify(execFile)

export interface GitContextResult {
  commits_today: number
  diff_lines: number
  first_commit_time: string | null
  repo_path: string | null
}

const COMMANDS: Record<string, string[]> = {
  commits_today: ['git', 'log', '--oneline', '--since=midnight'],
  diff_lines: ['git', 'diff', '--stat', 'HEAD'],
  first_commit_time: ['git', 'log', '--format=%ai', '--since=midnight'],
  repo_path: ['git', 'rev-parse', '--show-toplevel'],
}

export async function loadGitContext(cwd: string): Promise<GitContextResult> {
  const promises = Object.entries(COMMANDS).map(([key, cmd]) =>
    _runGitCmd(cmd, cwd).then(output => ({ key, output }))
  )

  const settled = await Promise.allSettled(promises)

  const results: GitContextResult = {
    commits_today: 0,
    diff_lines: 0,
    first_commit_time: null,
    repo_path: null,
  }

  for (const result of settled) {
    if (result.status === 'fulfilled') {
      const { key, output } = result.value
      if (output !== null) {
        if (key === 'commits_today') results.commits_today = _parseCommitsToday(output)
        else if (key === 'diff_lines') results.diff_lines = _parseDiffLines(output)
        else if (key === 'first_commit_time') results.first_commit_time = _parseFirstCommitTime(output)
        else if (key === 'repo_path') results.repo_path = _parseRepoPath(output)
      }
    }
    // result.status === 'rejected' → 静默跳过，使用默认值
  }

  return results
}

async function _runGitCmd(cmd: string[], cwd: string): Promise<string | null> {
  try {
    const { stdout } = await execFileAsync(cmd[0], cmd.slice(1), { cwd, timeout: 5000 })
    return stdout
  } catch {
    return null
  }
}

function _parseCommitsToday(output: string): number {
  return output.trim().split('\n').filter(line => line).length
}

function _parseDiffLines(output: string): number {
  if (!output.trim()) return 0

  let insertions = 0
  let deletions = 0

  const insMatch = output.match(/(\d+) insertion/)
  if (insMatch !== null) insertions = parseInt(insMatch[1], 10)

  const delMatch = output.match(/(\d+) deletion/)
  if (delMatch !== null) deletions = parseInt(delMatch[1], 10)

  return insertions + deletions
}

function _parseFirstCommitTime(output: string): string | null {
  const lines = output.trim().split('\n').filter(line => line)
  return lines[lines.length - 1] || null
}

function _parseRepoPath(output: string): string | null {
  const path = output.trim()
  return path || null
}
